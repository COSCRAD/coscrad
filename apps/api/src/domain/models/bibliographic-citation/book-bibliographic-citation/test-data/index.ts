export * from './build-book-bibliographic-citation-created';
