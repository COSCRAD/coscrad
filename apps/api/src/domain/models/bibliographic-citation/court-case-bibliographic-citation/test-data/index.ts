export * from './build-court-case-bibliographic-citation-created';
