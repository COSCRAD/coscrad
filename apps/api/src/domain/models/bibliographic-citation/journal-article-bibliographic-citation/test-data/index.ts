export * from './build-journal-article-bibliographic-citation-created';
