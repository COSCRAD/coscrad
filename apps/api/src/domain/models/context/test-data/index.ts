export * from './get-note-test-event-map';
