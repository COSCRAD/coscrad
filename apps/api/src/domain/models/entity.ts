import { PartialDTO } from 'apps/api/src/types/partial-dto';
import { EntityType } from '../types/entityType';
import { isNullOrUndefined } from '../utilities/validation/is-null-or-undefined';
import { EntityId } from './types/entity-id';
import { EntityCompositeIdentifier } from './types/entityCompositeIdentifier';

// TODO Extract to a separate location
const isValidStringWithLength = (input: unknown): input is string =>
  typeof input === 'string' && input.length > 0;

/**
 *
 * TODO move this near the corresponding type. Note that this extra layer is
 * meant to safe guard us in case we want stricter validation on an `entity id`.
 * Should we change the rule, we can update this in once place, instead of replacing
 * `isValidStringWithLength` everywhere.
 */
const isValidEntityId = (input: unknown): input is EntityId =>
  isValidStringWithLength(input);

export abstract class Entity {
  readonly id: EntityId;

  abstract readonly type: EntityType;

  readonly published: boolean;

  constructor(dto: unknown) {
    // TODO Validate DTO
    if (!this.#validateDTO(dto))
      throw new Error(`Invalid dto for an entity: ${JSON.stringify(dto)}`);

    this.id = dto.id;

    this.published = typeof dto.published === 'boolean' ? dto.published : false;
  }

  getCompositeIdentifier = (): EntityCompositeIdentifier => ({
    type: this.type,
    id: this.id,
  });

  /**
   * TODO consider moving this to a `Serializable` mixin. There may be cases
   * where we want this behaviour on a non domain-model class without the
   * inheritance baggage.
   *  */
  toDTO<TEntity extends Entity>(this: TEntity): PartialDTO<TEntity> {
    return JSON.parse(JSON.stringify(this));
  }

  /**
   * TODO develop a general pattern \ directory structure for validation \ validators.
   * One idea: Return a symbol Valid | Errors[] from a validator instead of a
   * boolean flag.
   *
   * We may break the validators out into separate files. Then we will validate
   * a DTO before passing it to the constructor in our `instance factory` methods
   * in the mapping layer for the repositories.
   *  */
  #validateDTO = (dto: unknown): dto is PartialDTO<Entity> => {
    const { id } = dto as PartialDTO<Entity>;

    // A new entity will have id auto assigned by db
    if (!isNullOrUndefined(id) && !isValidEntityId(id)) return false;

    return true;
  };
}
