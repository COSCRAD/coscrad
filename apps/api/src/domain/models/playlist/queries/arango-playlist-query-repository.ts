import { Observable } from 'rxjs';
import { InternalError } from '../../../../lib/errors/InternalError';
import { Maybe } from '../../../../lib/types/maybe';
import { isNotFound } from '../../../../lib/types/not-found';
import { ArangoConnectionProvider } from '../../../../persistence/database/arango-connection.provider';
import { ArangoDatabase } from '../../../../persistence/database/arango-database';
import { ArangoDatabaseForCollection } from '../../../../persistence/database/arango-database-for-collection';
import mapDatabaseDocumentToAggregateDTO from '../../../../persistence/database/utilities/mapDatabaseDocumentToAggregateDTO';
import mapEntityDTOToDatabaseDocument from '../../../../persistence/database/utilities/mapEntityDTOToDatabaseDocument';
import { PlaylistViewModel } from '../../../../queries/buildViewModelForResource/viewModels';
import { AggregateId } from '../../../types/AggregateId';
import { ArangoResourceQueryBuilder } from '../../term/repositories/arango-resource-query-builder';
import { IPlaylistQueryRepository } from './playlist-query-repository.interface';

export class ArangoPlaylistQueryRepository implements IPlaylistQueryRepository {
    private readonly database: ArangoDatabaseForCollection<PlaylistViewModel>;

    /**
     * We use this helper to achieve composition over inheritance.
     */
    private readonly baseResourceQueryBuilder: ArangoResourceQueryBuilder;

    constructor(
        // should we inject the database directly?
        arangoConnectionProvider: ArangoConnectionProvider
        // @Inject(COSCRAD_LOGGER_TOKEN)
        // readonly logger: ICoscradLogger
    ) {
        this.database = new ArangoDatabaseForCollection(
            new ArangoDatabase(arangoConnectionProvider.getConnection()),
            'playlist__VIEWS'
        );

        this.baseResourceQueryBuilder = new ArangoResourceQueryBuilder('playlist__VIEWS');
    }

    subscribeToUpdates(): Observable<{ data: { type: string } }> {
        throw new Error('Method not implemented.');
    }

    async create(view: PlaylistViewModel): Promise<void> {
        await this.database.create(mapEntityDTOToDatabaseDocument(view));
    }

    async createMany(views: PlaylistViewModel[]): Promise<void> {
        const documents = views.map(mapEntityDTOToDatabaseDocument);

        await this.database.createMany(documents);
    }

    delete(_id: AggregateId): Promise<void> {
        throw new Error('Method not implemented.');
    }

    async fetchById(id: AggregateId): Promise<Maybe<PlaylistViewModel>> {
        const result = await this.database.fetchById(id);

        if (isNotFound(result)) return result;

        const asView = mapDatabaseDocumentToAggregateDTO(result);

        return PlaylistViewModel.fromDto(asView);
    }

    async fetchMany(): Promise<PlaylistViewModel[]> {
        const result = await this.database.fetchMany();

        const asViews = result.map((doc) =>
            PlaylistViewModel.fromDto(mapDatabaseDocumentToAggregateDTO(doc))
        );

        return asViews;
    }

    allowUser(_id: AggregateId, _userId: AggregateId): Promise<void> {
        throw new Error('Method not implemented.');
    }

    async addAudioItem(playlistId: AggregateId, audioItemId: string) {
        const query = `
        FOR doc IN @@collectionName
        FILTER doc._key == @id
        FOR a IN audioItem__VIEWS
        FILTER a._key == @audioItemId
        LET nextEpisode = {
                name: a.name,
                mimeType: a.mimeType,
                lengthMilliseconds: a.lengthMilliseconds,
                mediaItemId: a.mediaItemId
            }
        UPDATE doc WITH {
            episodes: APPEND(doc.episodes,nextEpisode)
        } IN @@collectionName
         RETURN NEW
        `;

        const bindVars = {
            '@collectionName': 'playlist__VIEWS',
            id: playlistId,
            audioItemId,
        };

        const cursor = await this.database
            .query({
                query,
                bindVars,
            })
            .catch((reason) => {
                throw new InternalError(
                    `Failed to add audio item as episode to playlist via term query repository: ${reason}`
                );
            });

        const _new = await cursor.all();

        _new;
    }
}
