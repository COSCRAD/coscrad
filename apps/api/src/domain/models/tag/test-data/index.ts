export * from './get-tag-test-event-builders-map';
