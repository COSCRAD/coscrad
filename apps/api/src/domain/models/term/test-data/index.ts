export * from './build-test-term';
