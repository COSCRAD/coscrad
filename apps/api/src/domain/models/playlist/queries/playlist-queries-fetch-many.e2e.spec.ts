import {
    CoscradUserRole,
    HttpStatusCode,
    IIndexQueryResult,
    IPlayListViewModel,
} from '@coscrad/api-interfaces';
import { INestApplication } from '@nestjs/common';
import setUpIntegrationTest from '../../../../app/controllers/__tests__/setUpIntegrationTest';
import { ArangoConnectionProvider } from '../../../../persistence/database/arango-connection.provider';
import { ArangoDatabaseProvider } from '../../../../persistence/database/database.provider';
import generateDatabaseNameForTestSuite from '../../../../persistence/repositories/__tests__/generateDatabaseNameForTestSuite';
import TestRepositoryProvider from '../../../../persistence/repositories/__tests__/TestRepositoryProvider';
import { CoscradContributorViewModel } from '../../../../queries/buildViewModelForResource/viewModels/coscrad-contributor.view-model';
import {
    PlaylistEpisodeViewModel,
    PlaylistViewModel,
} from '../../../../queries/buildViewModelForResource/viewModels/playlist.view-model';
import { buildTestInstance } from '../../../../test-data/utilities';
import { buildMultilingualTextWithSingleItem } from '../../../common/build-multilingual-text-with-single-item';
import { assertQueryResult } from '../../__tests__';
import buildDummyUuid from '../../__tests__/utilities/buildDummyUuid';
import { AccessControlList } from '../../shared/access-control/access-control-list.entity';
import { CoscradUserWithGroups } from '../../user-management/user/entities/user/coscrad-user-with-groups';
import { CoscradUser } from '../../user-management/user/entities/user/coscrad-user.entity';
import { ArangoPlaylistQueryRepository } from './arango-playlist-query-repository';
import { IPlaylistQueryRepository } from './playlist-query-repository.interface';

const indexEndpoint = `/resources/playlists`;

const ordinaryUser = buildTestInstance(CoscradUser, {
    roles: [CoscradUserRole.viewer],
});

const publicEpisode = buildTestInstance(PlaylistEpisodeViewModel, {
    isPublished: true,
    name: buildMultilingualTextWithSingleItem('public episode'),
});

const privateEpisode = buildTestInstance(PlaylistEpisodeViewModel, {
    isPublished: false,
    name: buildMultilingualTextWithSingleItem('private episode'),
});

const unpublishedEpisodeWithAclAccessForOrdinaryUser = buildTestInstance(PlaylistEpisodeViewModel, {
    isPublished: false,
    accessControlList: new AccessControlList().allowUser(ordinaryUser.id),
    name: buildMultilingualTextWithSingleItem('unpublished episode with ACL access for viewer'),
});

const testContributor = buildTestInstance(CoscradContributorViewModel);

const contributions = [testContributor];

// + ALL
const publicPlaylist = buildTestInstance(PlaylistViewModel, {
    id: buildDummyUuid(1),
    isPublished: true,
    name: buildMultilingualTextWithSingleItem('public playlist'),
    episodes: [publicEpisode, privateEpisode, unpublishedEpisodeWithAclAccessForOrdinaryUser],
    contributions,
});

// - publicUser, - ordinaryUser, + admin
const privatePlaylist = buildTestInstance(PlaylistViewModel, {
    id: buildDummyUuid(2),
    isPublished: false,
    // no special access
    queryAccessControlList: new AccessControlList(),
    contributions,
});

// - publicUser, + ordinaryUser, + admin
const unpublishedPlaylistWithAclAccessForOrdinaryUser = buildTestInstance(PlaylistViewModel, {
    id: buildDummyUuid(3),
    isPublished: false,
    queryAccessControlList: new AccessControlList().allowUser(ordinaryUser.id),
    contributions,
});

const allPlaylists = [
    publicPlaylist,
    privatePlaylist,
    unpublishedPlaylistWithAclAccessForOrdinaryUser,
];

describe(`when querying for a playlist: fetch many (${indexEndpoint})`, () => {
    const testDatabaseName = generateDatabaseNameForTestSuite();

    let app: INestApplication;

    let testRepositoryProvider: TestRepositoryProvider;

    let databaseProvider: ArangoDatabaseProvider;

    let connectionProvider: ArangoConnectionProvider;

    let playlistQueryRepository: IPlaylistQueryRepository;

    const seedInitialState = async () => {
        await playlistQueryRepository.createMany(allPlaylists);
    };

    const setItUp = async (testUserWithGroups?: CoscradUserWithGroups) => {
        // TODO avoid using `setUpIntegrationTest` here
        ({ app, testRepositoryProvider, databaseProvider } = await setUpIntegrationTest(
            {
                ARANGO_DB_NAME: testDatabaseName,
            },
            {
                testUserWithGroups,
            }
            // no authenticated user
        ));

        connectionProvider = app.get(ArangoConnectionProvider);

        playlistQueryRepository = new ArangoPlaylistQueryRepository(connectionProvider);
    };

    beforeEach(async () => {
        await testRepositoryProvider.testSetup();
    });

    afterAll(async () => {
        databaseProvider.close();

        await app.close();
    });

    describe(`when the user is not authenticated (public user)`, () => {
        beforeAll(async () => {
            await setItUp();
        });

        it(`it should return only public vocabulary lists and episodes`, async () => {
            await assertQueryResult({
                app,
                endpoint: indexEndpoint,
                expectedStatus: HttpStatusCode.ok,
                seedInitialState,
                checkResponseBody: async (body: IIndexQueryResult<IPlayListViewModel>) => {
                    // The public user can only see 1 of the 3 playlists
                    expect(body.entities).toHaveLength(1);

                    // similiarly, the public user can see 1 of 3 episodes
                    expect(body.entities[0].episodes).toHaveLength(1);

                    expect(body.indexScopedActions).toHaveLength(0);

                    const entitiesWithAvailableActions = body.entities.filter(
                        ({ actions }) => Array.isArray(actions) && actions.length > 0
                    );

                    expect(entitiesWithAvailableActions).toEqual([]);
                },
            });
        });
    });

    describe(`when the user is authenticated as an ordinary user (viewer)`, () => {
        beforeAll(async () => {
            // TODO support user group based access control
            await setItUp(new CoscradUserWithGroups(ordinaryUser, []));
        });

        it(`it should return playlists and episodes that are public or have the user in the ACL`, async () => {
            await assertQueryResult({
                app,
                endpoint: indexEndpoint,
                expectedStatus: HttpStatusCode.ok,
                seedInitialState,
                checkResponseBody: async (body: IIndexQueryResult<IPlayListViewModel>) => {
                    // The viewer user can see the public playlist and the playlist with their user ID in the ACL
                    expect(body.entities).toHaveLength(2);

                    // The viewer user can see the public episode and the episode with their user ID in the ACL
                    expect(body.entities[0].episodes).toHaveLength(2);

                    expect(body.indexScopedActions).toHaveLength(0);

                    const entitiesWithAvailableActions = body.entities.filter(
                        ({ actions }) => Array.isArray(actions) && actions.length > 0
                    );

                    expect(entitiesWithAvailableActions).toEqual([]);
                },
            });
        });
    });

    describe(`when the user is authenticated as a COSCRAD admin`, () => {
        beforeAll(async () => {
            await setItUp(
                new CoscradUserWithGroups(
                    ordinaryUser.clone({
                        roles: [CoscradUserRole.superAdmin],
                    }),
                    []
                )
            );
        });

        it(`it should return playlists and episodes that are public or have the user in the ACL`, async () => {
            await assertQueryResult({
                app,
                endpoint: indexEndpoint,
                expectedStatus: HttpStatusCode.ok,
                seedInitialState,
                checkResponseBody: async (body: IIndexQueryResult<IPlayListViewModel>) => {
                    // an admin can see all playlists
                    expect(body.entities).toHaveLength(3);

                    // an admin can see all episodes
                    expect(body.entities[0].episodes).toHaveLength(3);

                    expect(body.indexScopedActions).not.toHaveLength(0);

                    const entitiesWithAvailableActions = body.entities.filter(
                        ({ actions }) => Array.isArray(actions) && actions.length > 0
                    );

                    expect(entitiesWithAvailableActions).toHaveLength(3);

                    /**
                     * We snapshot one and only one response as a test of the
                     * contract with the client. We choose the admin case
                     * because it includes `actions`.
                     */
                    expect(body).toMatchSnapshot();
                },
            });
        });
    });

    describe(`when the user is authenticated as a project admin`, () => {
        beforeAll(async () => {
            await setItUp(
                new CoscradUserWithGroups(
                    ordinaryUser.clone({
                        roles: [CoscradUserRole.projectAdmin],
                    }),
                    []
                )
            );
        });

        it(`it should return playlists and episodes that are public or have the user in the ACL`, async () => {
            await assertQueryResult({
                app,
                endpoint: indexEndpoint,
                expectedStatus: HttpStatusCode.ok,
                seedInitialState,
                checkResponseBody: async (body: IIndexQueryResult<IPlayListViewModel>) => {
                    // an admin can see all playlists
                    expect(body.entities).toHaveLength(3);

                    // an admin can see all episodes
                    expect(body.entities[0].episodes).toHaveLength(3);

                    expect(body.indexScopedActions).not.toHaveLength(0);

                    const entitiesWithAvailableActions = body.entities.filter(
                        ({ actions }) => Array.isArray(actions) && actions.length > 0
                    );

                    expect(entitiesWithAvailableActions).toHaveLength(3);
                },
            });
        });
    });
});
