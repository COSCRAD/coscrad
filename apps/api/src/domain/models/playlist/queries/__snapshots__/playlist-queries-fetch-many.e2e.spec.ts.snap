// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when querying for a playlist: fetch many (/resources/playlists) when the user is authenticated as a COSCRAD admin it should return playlists and episodes that are public or have the user in the ACL 1`] = `
{
  "entities": [
    {
      "actions": [
        {
          "description": "Make a resource visible to the public",
          "form": {
            "fields": [],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
                "type": "playlist",
              },
            },
          },
          "label": "Publish Resource",
          "type": "PUBLISH_RESOURCE",
        },
        {
          "description": "Assign a tag to a resource or note",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource or note being tagged",
                "label": "Tagged Member's Composite Identifier",
                "name": "taggedMemberCompositeIdentifier",
                "type": "JSON_INPUT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
                "type": "playlist",
              },
            },
          },
          "label": "Tag Resource or Note",
          "type": "TAG_RESOURCE_OR_NOTE",
        },
        {
          "description": "creates a note about this particular resource",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource about which we are making a note",
                "label": "CompositeIdentifier",
                "name": "resourceCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the note",
                "label": "text",
                "name": "text",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are writing the note",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
                "type": "playlist",
              },
            },
          },
          "label": "Create Note",
          "type": "CREATE_NOTE_ABOUT_RESOURCE",
        },
      ],
      "contributions": [
        {
          "contributorIds": [],
          "date": {
            "day": 26,
            "month": "09",
            "year": 2022,
          },
          "statement": "Watchamacallit fabricated by: Willy Wankie",
          "type": "WHATCHAMACALLIT_FABRICATED",
        },
      ],
      "episodes": [
        {
          "isPublished": true,
          "mediaItemUrl": "https:www.coscrad.org/api/resources/mediaItems/download/9b1deb4d-3b7d-4bad-9bdd-2b0d7b100567",
          "name": {
            "items": [
              {
                "languageCode": "en",
                "role": "original",
                "text": "public episode",
              },
            ],
          },
        },
        {
          "isPublished": false,
          "mediaItemUrl": "https:www.coscrad.org/api/resources/mediaItems/download/9b1deb4d-3b7d-4bad-9bdd-2b0d7b100567",
          "name": {
            "items": [
              {
                "languageCode": "en",
                "role": "original",
                "text": "private episode",
              },
            ],
          },
        },
        {
          "isPublished": false,
          "mediaItemUrl": "https:www.coscrad.org/api/resources/mediaItems/download/9b1deb4d-3b7d-4bad-9bdd-2b0d7b100567",
          "name": {
            "items": [
              {
                "languageCode": "en",
                "role": "original",
                "text": "unpublished episode with ACL access for viewer",
              },
            ],
          },
        },
      ],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100001",
      "isPublished": true,
      "name": {
        "items": [
          {
            "languageCode": "en",
            "role": "original",
            "text": "public playlist",
          },
        ],
      },
    },
    {
      "actions": [
        {
          "description": "Make a resource visible to the public",
          "form": {
            "fields": [],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100002",
                "type": "playlist",
              },
            },
          },
          "label": "Publish Resource",
          "type": "PUBLISH_RESOURCE",
        },
        {
          "description": "Assign a tag to a resource or note",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource or note being tagged",
                "label": "Tagged Member's Composite Identifier",
                "name": "taggedMemberCompositeIdentifier",
                "type": "JSON_INPUT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100002",
                "type": "playlist",
              },
            },
          },
          "label": "Tag Resource or Note",
          "type": "TAG_RESOURCE_OR_NOTE",
        },
        {
          "description": "creates a note about this particular resource",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource about which we are making a note",
                "label": "CompositeIdentifier",
                "name": "resourceCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the note",
                "label": "text",
                "name": "text",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are writing the note",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100002",
                "type": "playlist",
              },
            },
          },
          "label": "Create Note",
          "type": "CREATE_NOTE_ABOUT_RESOURCE",
        },
      ],
      "contributions": [
        {
          "contributorIds": [],
          "date": {
            "day": 26,
            "month": "09",
            "year": 2022,
          },
          "statement": "Watchamacallit fabricated by: Willy Wankie",
          "type": "WHATCHAMACALLIT_FABRICATED",
        },
      ],
      "episodes": [],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100002",
      "isPublished": false,
      "name": {
        "items": [
          {
            "languageCode": "en",
            "role": "original",
            "text": "Metal Mondays",
          },
        ],
      },
    },
    {
      "actions": [
        {
          "description": "Make a resource visible to the public",
          "form": {
            "fields": [],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100003",
                "type": "playlist",
              },
            },
          },
          "label": "Publish Resource",
          "type": "PUBLISH_RESOURCE",
        },
        {
          "description": "Assign a tag to a resource or note",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource or note being tagged",
                "label": "Tagged Member's Composite Identifier",
                "name": "taggedMemberCompositeIdentifier",
                "type": "JSON_INPUT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100003",
                "type": "playlist",
              },
            },
          },
          "label": "Tag Resource or Note",
          "type": "TAG_RESOURCE_OR_NOTE",
        },
        {
          "description": "creates a note about this particular resource",
          "form": {
            "fields": [
              {
                "constraints": [],
                "description": "system-wide unique identifier for the resource about which we are making a note",
                "label": "CompositeIdentifier",
                "name": "resourceCompositeIdentifier",
                "type": "JSON_INPUT",
              },
              {},
              {
                "constraints": [
                  {
                    "message": "must be a non-empty string",
                    "name": "non-empty string",
                  },
                  {
                    "message": "must be a defined value",
                    "name": "defined value",
                  },
                ],
                "description": "text for the note",
                "label": "text",
                "name": "text",
                "type": "TEXT_FIELD",
              },
              {
                "constraints": [
                  {
                    "message": "Must be a valid \${propertyLabel}",
                    "name": "IS_ENUM",
                  },
                  {
                    "message": "Required",
                    "name": "defined value",
                  },
                ],
                "description": "the language in which you are writing the note",
                "label": "language code",
                "name": "languageCode",
                "options": [
                  {
                    "display": "Chilcotin",
                    "value": "clc",
                  },
                  {
                    "display": "Haida",
                    "value": "hai",
                  },
                  {
                    "display": "English",
                    "value": "en",
                  },
                  {
                    "display": "French",
                    "value": "fra",
                  },
                  {
                    "display": "Chinook",
                    "value": "chn",
                  },
                  {
                    "display": "Zapotec",
                    "value": "zap",
                  },
                  {
                    "display": "Spanish",
                    "value": "spa",
                  },
                ],
                "type": "STATIC_SELECT",
              },
            ],
            "prepopulatedFields": {
              "aggregateCompositeIdentifier": {
                "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100003",
                "type": "playlist",
              },
            },
          },
          "label": "Create Note",
          "type": "CREATE_NOTE_ABOUT_RESOURCE",
        },
      ],
      "contributions": [
        {
          "contributorIds": [],
          "date": {
            "day": 26,
            "month": "09",
            "year": 2022,
          },
          "statement": "Watchamacallit fabricated by: Willy Wankie",
          "type": "WHATCHAMACALLIT_FABRICATED",
        },
      ],
      "episodes": [],
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100003",
      "isPublished": false,
      "name": {
        "items": [
          {
            "languageCode": "en",
            "role": "original",
            "text": "Metal Mondays",
          },
        ],
      },
    },
  ],
  "indexScopedActions": [
    {
      "description": "Creates a new playlist",
      "form": {
        "fields": [
          {
            "constraints": [
              {
                "message": "must be a non-empty string",
                "name": "non-empty string",
              },
              {
                "message": "must be a defined value",
                "name": "defined value",
              },
            ],
            "description": "name of the playlist",
            "label": "name",
            "name": "name",
            "type": "TEXT_FIELD",
          },
          {
            "constraints": [
              {
                "message": "Must be a valid \${propertyLabel}",
                "name": "IS_ENUM",
              },
              {
                "message": "Required",
                "name": "defined value",
              },
            ],
            "description": "the language in which you are naming this playlist",
            "label": "language code",
            "name": "languageCodeForName",
            "options": [
              {
                "display": "Chilcotin",
                "value": "clc",
              },
              {
                "display": "Haida",
                "value": "hai",
              },
              {
                "display": "English",
                "value": "en",
              },
              {
                "display": "French",
                "value": "fra",
              },
              {
                "display": "Chinook",
                "value": "chn",
              },
              {
                "display": "Zapotec",
                "value": "zap",
              },
              {
                "display": "Spanish",
                "value": "spa",
              },
            ],
            "type": "STATIC_SELECT",
          },
        ],
        "prepopulatedFields": {},
      },
      "label": "Create Playlist",
      "type": "CREATE_PLAYLIST",
    },
  ],
}
`;
